src/theory/datatypes/inference_manager.cpp

```diff
@@ -0,0 +1,184 @@
+/*********************                                                        */
+/*! \file inference_manager.cpp
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Andrew Reynolds
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2020 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief Datatypes inference manager
+ **/
+
+#include "theory/datatypes/inference_manager.h"
+
+#include "expr/dtype.h"
+#include "options/datatypes_options.h"
+#include "theory/theory.h"
+
+using namespace CVC4::kind;
+
+namespace CVC4 {
+namespace theory {
+namespace datatypes {
+
+InferenceManager::InferenceManager(Theory& t,
+                                   TheoryState& state,
+                                   ProofNodeManager* pnm)
+    : InferenceManagerBuffered(t, state, pnm),
+      d_lemmasSent(t.getSatContext()),
+      d_addedLemma(false),
+      d_addedFact(false)
+{
+  d_true = NodeManager::currentNM()->mkConst(true);
+}
+
+void InferenceManager::reset()
+{
+  d_addedLemma = false;
+  d_addedFact = false;
+}
+
+bool InferenceManager::mustCommunicateFact(Node n, Node exp) const
+{
+  Trace("dt-lemma-debug") << "Compute for " << exp << " => " << n << std::endl;
+  bool addLemma = false;
+  if (options::dtInferAsLemmas() && exp != d_true)
+  {
+    addLemma = true;
+  }
+  else if (n.getKind() == EQUAL)
+  {
+    TypeNode tn = n[0].getType();
+    if (!tn.isDatatype())
+    {
+      addLemma = true;
+    }
+    else
+    {
+      const DType& dt = tn.getDType();
+      addLemma = dt.involvesExternalType();
+    }
+  }
+  else if (n.getKind() == LEQ || n.getKind() == OR)
+  {
+    addLemma = true;
+  }
+  if (addLemma)
+  {
+    Trace("dt-lemma-debug") << "Communicate " << n << std::endl;
+    return true;
+  }
+  Trace("dt-lemma-debug") << "Do not need to communicate " << n << std::endl;
+  return false;
+}
+
+void InferenceManager::process()
+{
+  // process pending lemmas, used infrequently, only for definitional lemmas
+  doPendingLemmas();
```

Comment:
If `numPendingLemmas() > 0`, we should set `d_addedLemmas = true` here.

Reply:
This is now standardized at a higher level.